security/pkg/k8s/chiron/controller.go

```diff
@@ -0,0 +1,591 @@
+			gracePeriodRatio, recommendedMinGracePeriodRatio, recommendedMaxGracePeriodRatio)
+	}
+
+	c := &WebhookController{
+		gracePeriodRatio:           gracePeriodRatio,
+		minGracePeriod:             minGracePeriod,
+		k8sCaCertFile:              k8sCaCertFile,
+		k8sClient:                  k8sClient,
+		core:                       core,
+		certClient:                 certClient,
+		namespace:                  nameSpace,
+		mutatingWebhookConfigFiles: mutatingWebhookConfigFiles,
+		mutatingWebhookConfigNames: mutatingWebhookConfigNames,
+	}
+
+	// read CA cert at the beginning of launching the controller and when the CA cert changes.
+	caCert, err := readCACert(k8sCaCertFile)
+	if err != nil {
+		log.Errorf("failed to read CA certificate: %v", err)
+		return nil, err
+	}
+	c.setCurCACert(caCert)
+
+	namespaces := []string{nameSpace}
+
+	istioSecretSelector := fields.SelectorFromSet(map[string]string{"type": IstioSecretType}).String()
+	scrtLW := listwatch.MultiNamespaceListerWatcher(namespaces, func(namespace string) cache.ListerWatcher {
+		return &cache.ListWatch{
+			ListFunc: func(options metav1.ListOptions) (runtime.Object, error) {
+				options.FieldSelector = istioSecretSelector
+				return core.Secrets(namespace).List(options)
+			},
+			WatchFunc: func(options metav1.ListOptions) (watch.Interface, error) {
+				options.FieldSelector = istioSecretSelector
+				return core.Secrets(namespace).Watch(options)
+			},
+		}
+	})
+	// The certificate rotation is handled by scrtUpdated().
+	c.scrtStore, c.scrtController =
+		cache.NewInformer(scrtLW, &v1.Secret{}, secretResyncPeriod, cache.ResourceEventHandlerFuncs{
+			DeleteFunc: c.scrtDeleted,
+			UpdateFunc: c.scrtUpdated,
+		})
+
+	// Create a watcher for the CA certificate such that when
+	// the CA certificate changes, the webhook certificates are regenerated and
+	// the CA bundles in webhook configurations get updated.
+	watcher, err := fsnotify.NewWatcher()
+	if err != nil {
+		return nil, err
+	}
+	// Watch the parent directory of the target files so we can catch
+	// symlink updates of k8s ConfigMaps volumes.
+	// The files watched include the CA certificate file and the muatingwebhookconfiguration file,
+	// which is a ConfigMap file mount.
+	// In the prototype, only the first mutatingwebhookconfiguration is watched.
+	for _, file := range []string{k8sCaCertFile, mutatingWebhookConfigFiles[0]} {
+		watchDir, _ := filepath.Split(file)
+		if err := watcher.Watch(watchDir); err != nil {
+			return nil, fmt.Errorf("could not watch %v: %v", file, err)
+		}
+	}
+	c.ConfigWatcher = watcher
+
+	return c, nil
+}
+
+// Run starts the WebhookController until stopCh is notified.
+func (wc *WebhookController) Run(stopCh chan struct{}) {
+	// Create secrets containing certificates for webhooks
+	for _, scrtName := range WebhookSecretNames {
+		wc.upsertSecret(scrtName, wc.namespace)
+	}
+
+	host := fmt.Sprintf("%s.%s", WebhookServiceNames[0], wc.namespace)
```

Comment:
Please create a loop to check and create webhooks for all the WebhookServiceNames.

Reply:
Because different webhook services have different event handlers, which are listed in the switch case handlers as follows, currently only one mutating webhook (i.e., Sidecar Injector) and one validating webhook (Galley) are supported.
	select {
		case <-timerMutateWhFile:
                ....
		case <-timerValidateWhFile:
                ....
         	case event := <-wc.MutatingWebhookFileWatcher.Event:
                ....
		case err := <-wc.MutatingWebhookFileWatcher.Error:
	        .....
		case event := <-wc.ValidatingWebhookFileWatcher.Event:
                ....
        	case err := <-wc.ValidatingWebhookFileWatcher.Error:
                ....
		case <-mutatingWebhookChangedCh:
                .....
		case <-validatingWebhookChangedCh:
                ......
       }